clippy_lints/src/loops/never_loop.rs

```diff
@@ -1,13 +1,37 @@
 use super::NEVER_LOOP;
-use clippy_utils::diagnostics::span_lint;
-use rustc_hir::{Block, Expr, ExprKind, HirId, InlineAsmOperand, Stmt, StmtKind};
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::higher;
+use clippy_utils::source::{snippet, snippet_opt};
+use clippy_utils::ty::has_iter_method;
+use rustc_ast::util::parser::PREC_POSTFIX;
+use rustc_errors::Applicability;
+use rustc_hir::{is_range_literal, Block, Expr, ExprKind, HirId, InlineAsmOperand, LoopSource, Pat, Stmt, StmtKind};
 use rustc_lint::LateContext;
 use std::iter::{once, Iterator};
 
 pub(super) fn check(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
-    if let ExprKind::Loop(block, _, _, _) = expr.kind {
+    if let ExprKind::Loop(block, _, source, _) = expr.kind {
         match never_loop_block(block, expr.hir_id) {
-            NeverLoopResult::AlwaysBreak => span_lint(cx, NEVER_LOOP, expr.span, "this loop never actually loops"),
+            NeverLoopResult::AlwaysBreak => {
+                span_lint_and_then(cx, NEVER_LOOP, expr.span, "this loop never actually loops", |diag| {
+                    if_chain! {
+                        if let LoopSource::ForLoop = source;
+                        let parent_arm = cx.tcx.hir().get_parent_node(expr.hir_id);
+                        let parent_match = cx.tcx.hir().expect_expr(cx.tcx.hir().get_parent_node(parent_arm));
+                        if let Some((pat, iterator, _, _)) = higher::for_loop(parent_match);
+                        if let Some(sugg) = for_to_if_let_sugg(cx, iterator, pat);
+                        then {
+                            diag.span_suggestion_verbose(
+                                cx.tcx.sess.source_map().guess_head_span(expr.span),
```

Comment:
A "guess" shouldn't be necessary. Try `span.with_hi(iterator.span)` where `span` is the last element returned from `higher::for_loop`.